test/lit/passes/type-ssa_and_merging.wast

```diff
@@ -0,0 +1,143 @@
+  ;; NOP-NEXT:  (local.set $0
+  ;; NOP-NEXT:   (struct.new $A
+  ;; NOP-NEXT:    (i32.const 1337)
+  ;; NOP-NEXT:   )
+  ;; NOP-NEXT:  )
+  ;; NOP-NEXT:  (call $escape
+  ;; NOP-NEXT:   (local.tee $1
+  ;; NOP-NEXT:    (struct.new $A
+  ;; NOP-NEXT:     (i32.const 42)
+  ;; NOP-NEXT:    )
+  ;; NOP-NEXT:   )
+  ;; NOP-NEXT:  )
+  ;; NOP-NEXT:  (call $escape
+  ;; NOP-NEXT:   (local.get $0)
+  ;; NOP-NEXT:  )
+  ;; NOP-NEXT:  (struct.set $A 0
+  ;; NOP-NEXT:   (local.get $1)
+  ;; NOP-NEXT:   (i32.const 100)
+  ;; NOP-NEXT:  )
+  ;; NOP-NEXT:  (struct.set $A 0
+  ;; NOP-NEXT:   (local.get $0)
+  ;; NOP-NEXT:   (i32.const 1337)
+  ;; NOP-NEXT:  )
+  ;; NOP-NEXT:  (i32.add
+  ;; NOP-NEXT:   (struct.get $A 0
+  ;; NOP-NEXT:    (local.get $1)
+  ;; NOP-NEXT:   )
+  ;; NOP-NEXT:   (struct.get $A 0
+  ;; NOP-NEXT:    (local.get $0)
+  ;; NOP-NEXT:   )
+  ;; NOP-NEXT:  )
+  ;; NOP-NEXT: )
+  ;; YES:      (export "foo" (func $foo))
+
+  ;; YES:      (func $foo (type $none_=>_i32) (result i32)
+  ;; YES-NEXT:  (local $0 (ref $A))
+  ;; YES-NEXT:  (local $1 (ref $A))
+  ;; YES-NEXT:  (local.set $0
+  ;; YES-NEXT:   (struct.new $A
+  ;; YES-NEXT:    (i32.const 1337)
+  ;; YES-NEXT:   )
+  ;; YES-NEXT:  )
+  ;; YES-NEXT:  (call $escape
+  ;; YES-NEXT:   (local.tee $1
+  ;; YES-NEXT:    (struct.new $A
+  ;; YES-NEXT:     (i32.const 42)
+  ;; YES-NEXT:    )
+  ;; YES-NEXT:   )
+  ;; YES-NEXT:  )
+  ;; YES-NEXT:  (call $escape
+  ;; YES-NEXT:   (local.get $0)
+  ;; YES-NEXT:  )
+  ;; YES-NEXT:  (struct.set $A 0
+  ;; YES-NEXT:   (local.get $1)
+  ;; YES-NEXT:   (i32.const 100)
+  ;; YES-NEXT:  )
+  ;; YES-NEXT:  (struct.set $A 0
+  ;; YES-NEXT:   (local.get $0)
+  ;; YES-NEXT:   (i32.const 1337)
+  ;; YES-NEXT:  )
+  ;; YES-NEXT:  (i32.add
+  ;; YES-NEXT:   (struct.get $A 0
+  ;; YES-NEXT:    (local.get $1)
+  ;; YES-NEXT:   )
+  ;; YES-NEXT:   (i32.const 1337)
+  ;; YES-NEXT:  )
+  ;; YES-NEXT: )
+  (func $foo (export "foo") (result i32)
+    (local $a-42 (ref null $A))
+    (local $a-1337 (ref null $A))
+
+    ;; Two instances of $a. Both escape, so we can't just do escape analysis
+    ;; here, and the field is mutable.
+    (local.set $a-42
+      (struct.new $A
+        (i32.const 42)
+      )
+    )
+    (local.set $a-1337
+      (struct.new $A
+        (i32.const 1337)
+      )
+    )
+    (call $escape
+      (local.get $a-42)
+    )
+    (call $escape
+      (local.get $a-1337)
+    )
+
+    ;; Write some values back to them. The first gets a different value than
+    ;; before.
+    (struct.set $A 0
+      (local.get $a-42)
+      (i32.const 100)
+    )
+    (struct.set $A 0
+      (local.get $a-1337)
+      (i32.const 1337)
+    )
+
+    ;; We can infer 1337 in the second local: that type, after SSA, always gets
+    ;; written the same constant.
+    (i32.add
+      (struct.get $A 0
+        (local.get $a-42)
+      )
+      (struct.get $A 0       ;; in YES this will become a constant 1337
+        (local.get $a-1337)
+      )
+    )
```

Comment:
Shouldn't precompute-propagate be able to optimize both of these already? I would have expected the interesting cases for this pair of optimizations to be interprocedural.

Reply:
`precompute-propagate` can reason about immutable fields, but not mutable ones (a mutable field that escapes cannot be inferred on; and that pass doesn't look for escaping as we assume that if there is no escaping then escape analysis would have removed the entire allocation anyhow).

Reply:
(But we could do load-store forwarding in specific intervals where no escape exists, even if escapes exist elsewhere. That would go under the topic of dead store elimination, which we might want to return to eventually.)

Reply:
Abstract interpretation could also handle this.

It may be more future-proof to make this test cover multiple functions so that it still tests what it means to test in the face of future optimizations.

Reply:
How do you suggest to split it? The problem I run into is that we need to have a direct path from the `struct.new`s, which get new types, to the uses of them. So we can't just split that existing function into 2 at any point - we can't just pass a general thing of the original single type, we need to pass that type on two different paths that each originate from exactly one `struct.new`.

Or we could split into multiple functions called by the main one, but then they'd get inlined.

Reply:
I have something like this in mind:
```
(func $main
  (call $get-a-1
    (struct.new $A (i32.const 42))
  )

  (call $get-a-2
    (struct.new $A (i32.const 1337))
  )
)

(func $get-a-1 (param (ref $A))
  (drop (struct.get $A 0 (local.get 0)))
)

(func $get-a-2 (param (ref $A)) ;; This could be generated by monomorphization
  (drop (struct.get $A 0 (local.get 0)))
)
```

I guess that's covered by

> Or we could split into multiple functions called by the main one, but then they'd get inlined.

Could we add loops or import calls to defeat inlining?